### PR TITLE
Send bus notifications if first leg is bus

### PR DESCRIPTION
**Pull request by br648**
### Checklist

- [x] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [x] Any modified or new methods or classes have helpful JavaDoc and code is thoroughly commented
- [x] The description lists all applicable issues this PR seeks to resolve
- [x] The description lists any configuration setting(s) that differ from the default settings
- [x] All tests and CI builds passing

### Description

If the trip starts with a bus leg notify bus op.

### CODE REVIEW

**Comment by br648** [comment]
@binh-dam-ibigroup the test that is failing also fails under dev. I think it is a time/locale issue. I bet it will work when I re-run it tomorrow morning!

Failed test has this comment: `// Time is in US Pacific time zone (instead of US Eastern) by configuration for other E2E tests.`

File: src/main/java/org/opentripplanner/middleware/triptracker/TravelerLocator.java [hunk 1]
```diff
@@ -157,6 +154,32 @@ public static TripInstruction alignTravelerToTrip(
...
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Given the traveler's position and leg type, check if bus notification should be sent.
+     */
+    public static boolean shouldNotifyBusOperator(TravelerPosition travelerPosition, boolean isStartOfTrip) {
+        return (isStartOfTrip)
```
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/triptracker/TravelerLocator.java]: Extract a variable for the leg to pass to `isBusLeg` and `isWithinOperationalNotifyWindow` depending on `isStartOfTrip`, and remove line 180.

**Review by binh-dam-ibigroup — CHANGES_REQUESTED** [review]
I think you need to add a test case for an itinerary which first leg is a bus leg. I am seeing a side effect where, when you start tracking with coordinates near the bus stop, it thinks you are already on the bus.

File: src/main/java/org/opentripplanner/middleware/triptracker/TravelerLocator.java [hunk 2]
```diff
@@ -136,10 +136,7 @@ public static TripInstruction alignTravelerToTrip(
         Locale locale = travelerPosition.locale;
 
         if (isApproachingEndOfLeg(travelerPosition)) {
```
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/triptracker/TravelerLocator.java:176]: If this is at the start of a trip where the first leg is transit, there is no approaching the end of the leg.

Comment (br648) [reply src/main/java/org/opentripplanner/middleware/triptracker/TravelerLocator.java:176]: @binh-dam-ibigroup  If possible can you provide a trip which starts with a transit leg? I'm going to edit an existing trip (walk-to-bus-transition.json) to test with, but would prefer a real-world exampl

Comment (br648) [reply src/main/java/org/opentripplanner/middleware/triptracker/TravelerLocator.java:176]: @binh-dam-ibigroup this has grown a bit from my initial take on it! I think this covers the what is needed.

**Comment by binh-dam-ibigroup** [comment]
@br648 One of the E2E tests is failing and outputting an in incorrect instruction...

File: src/main/java/org/opentripplanner/middleware/triptracker/ManageTripTracking.java [hunk 3]
```diff
@@ -145,19 +151,40 @@ private static EndTrackingResponse completeJourney(TripTrackingData tripData, bo
...
 
-        // Provide response.
         return new EndTrackingResponse(
             TripInstruction.NO_INSTRUCTION,
             TripStatus.ENDED.name()
         );
+    }
 
+    /**
+     * Cancel bus notifications which will not be fulfilled.
```
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/triptracker/ManageTripTracking.java]:
```suggestion
     * Cancel bus notifications which are no longer needed/relevant.
```

**Review by binh-dam-ibigroup — COMMENTED** [review]
Just one comment tweak, but I am worried about the failing E2E test right now.

**Comment by br648** [comment]
> @br648 One of the E2E tests is failing and outputting an in incorrect instruction...

@binh-dam-ibigroup The same test is failing under dev and is not specific to this PR. The CI incorrect instruction is different to the one I get locally. Will see if I can fix.

File: src/main/java/org/opentripplanner/middleware/triptracker/TravelerLocator.java [hunk 4]
```diff
@@ -130,16 +129,12 @@ private static TripInstruction getBackOnTrack(
...
+        boolean isStartOfTrip
     ) {
         Locale locale = travelerPosition.locale;
 
         if (isApproachingEndOfLeg(travelerPosition)) {
-            if (isBusLeg(travelerPosition.nextLeg) && isWithinOperationalNotifyWindow(travelerPosition)) {
-                BusOperatorActions
-                    .getDefault()
-                    .handleSendNotificationAction(tripStatus, travelerPosition);
+            if (sendBusNotification(travelerPosition, isStartOfTrip)) {
```
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/triptracker/TravelerLocator.java]: You might have misinterpreted `isStartOfTrip` when using with `sendBusNotification`, and that is what is causing the failing trip tracking E2E test. I think `true` means that this is the first tracking location that is sent (the user just activated live tracking). I think you meant to pass to `sendBusNotification` a boolean indicating whether a transit leg is the first leg of the trip.

Comment (br648) [reply src/main/java/org/opentripplanner/middleware/triptracker/TravelerLocator.java]: @binh-dam-ibigroup I have updated, I hope it meets expectations.

File: src/main/java/org/opentripplanner/middleware/triptracker/TravelerLocator.java [hunk 5]
```diff
@@ -57,14 +58,14 @@ public static String getInstruction(
...
-                TripInstruction tripInstruction = alignTravelerToTrip(travelerPosition, isStartOfTrip, tripStatus);
+                TripInstruction tripInstruction = alignTravelerToTrip(travelerPosition, isStartOfTrip);
                 if (tripInstruction != null) {
                     return tripInstruction.build();
                 }
             }
 
             if (tripStatus.equals(TripStatus.DEVIATED)) {
-                TripInstruction tripInstruction = getBackOnTrack(travelerPosition, isStartOfTrip, tripStatus);
+                TripInstruction tripInstruction = getBackOnTrack(travelerPosition, isStartOfTrip);
```
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/triptracker/TravelerLocator.java:69]: Bonus points question: I didn't see an equivalent to "Head to <departure bus stop>" on an itinerary where the first leg is bus and tracking is started at a location away from the bus stop.

**Review by binh-dam-ibigroup — APPROVED** [review]
Much better!

File: src/main/java/org/opentripplanner/middleware/triptracker/TravelerLocator.java (hunk 5 above)
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/triptracker/TravelerLocator.java:69]: @binh-dam-ibigroup Update here: https://github.com/ibi-group/otp-middleware/pull/264/commits/eeea56917235f4b7c431e1b58937d23c242c334f

**Review by JymDyerIBI — APPROVED** [review]
👍  This looks good.

File: src/main/java/org/opentripplanner/middleware/triptracker/TravelerLocator.java (hunk 5 above)
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/triptracker/TravelerLocator.java:69]: @binh-dam-ibigroup I will hold off merging until you have "approved" this addition.

**Pull request merged**